From 7dc69191376dcf8e64f46e9f95e563da3940b096 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 27 Sep 2023 11:48:37 +0200
Subject: [PATCH 366/841] feat(stm32mp2): add add a function to set ICN QOS

By default this function does nothing, the reset values for SYSCFG_ICN*
registers fit ST boards needs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 3275923753724a7d93d1d23b6caea38284e87b4d feat(stm32mp2): add add a function to set ICN QOS]
Change-Id: I3aa16504bd47122515500a77d773023231d84a5a
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Signed-off-by: Antonio Borneo <antonio.borneo@foss.st.com>
Signed-off-by: Loic PALLARDY <loic.pallardy@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/329908
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/bl2_plat_setup.c           |  3 ++
 plat/st/stm32mp2/include/stm32mp2_private.h |  2 +
 plat/st/stm32mp2/stm32mp2_syscfg.c          | 41 +++++++++++++++++++++
 3 files changed, 46 insertions(+)

diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 60f79446a..eba386eda 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -177,6 +177,9 @@ void bl2_platform_setup(void)
 		ERROR("DDR mapping: error %d\n", ret);
 		panic();
 	}
+
+	/* Set QOS ICN priority */
+	stm32mp2_syscfg_set_icn_qos();
 }
 
 static void reset_backup_domain(void)
diff --git a/plat/st/stm32mp2/include/stm32mp2_private.h b/plat/st/stm32mp2/include/stm32mp2_private.h
index 9300be8c0..ba317e887 100644
--- a/plat/st/stm32mp2/include/stm32mp2_private.h
+++ b/plat/st/stm32mp2/include/stm32mp2_private.h
@@ -35,6 +35,8 @@ enum syscfg_io_ids {
 
 void stm32mp2_syscfg_enable_io_compensation(enum syscfg_io_ids id);
 
+void stm32mp2_syscfg_set_icn_qos(void);
+
 int stm32mp2_pwr_init_io_domains(void);
 
 /* Get RISAF platform instance ID from peripheral IO memory base address */
diff --git a/plat/st/stm32mp2/stm32mp2_syscfg.c b/plat/st/stm32mp2/stm32mp2_syscfg.c
index bae68aa22..8ff8173aa 100644
--- a/plat/st/stm32mp2/stm32mp2_syscfg.c
+++ b/plat/st/stm32mp2/stm32mp2_syscfg.c
@@ -78,6 +78,21 @@
 #define SYSCFG_DLYBOS_TAPSEL_NB		33U
 #define SYSCFG_DLYBOS_CMD_NB		24U
 
+/* ICN registers */
+#define SYSCFG_ICNQPCR1			U(0x2000)
+#define SYSCFG_ICNQPCR2			U(0x2004)
+#define SYSCFG_ICNEWRCR			U(0x2008)
+#define SYSCFG_ICNCGCR			U(0x200C)
+#define SYSCFG_ICNGPUBWLCR		U(0x2010)
+#define SYSCFG_ICNE2EBWRCR		U(0x2014)
+#define SYSCFG_ICNPCIBWLCR		U(0x201C)
+#define SYSCFG_ICNETHBWLCR		U(0x2020)
+#define SYSCFG_ICNUSB3BWLCR		U(0x2024)
+#define SYSCFG_ICNCPU1BWLCR		U(0x2028)
+#define SYSCFG_ICNLTDCBWLCR		U(0x202C)
+#define SYSCFG_ICNDCMIPPBWLCR		U(0x2030)
+#define SYSCFG_ICNVDEBWLCR		U(0x2034)
+
 struct syscfg_tap_window {
 	uint8_t end;
 	uint8_t length;
@@ -482,3 +497,29 @@ int stm32mp2_syscfg_dlyb_init(uint8_t bank, bool bypass_mode,
 	return 0;
 }
 #endif
+
+/*
+ * @brief  Set ICN QOS priority per master.
+ *	   By default, SYSCFG_ICN* register keep their reset values
+ * @retval 0.
+ */
+void stm32mp2_syscfg_set_icn_qos(void)
+{
+/*
+ *	Reset values, un-comment and change if required
+ *
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNQPCR1, U(0xCCCBBB3E));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNQPCR2, U(0x2449865A));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNEWRCR, U(0x0));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNCGCR, U(0x0));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNGPUBWLCR, U(0x0000400B));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNE2EBWRCR, U(0x00000052));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNPCIBWLCR, U(0x00002002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNETHBWLCR, U(0x00001002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNUSB3BWLCR, U(0x00004002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNCPU1BWLCR, U(0x00002002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNLTDCBWLCR, U(0x00008002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNDCMIPPBWLCR, U(0x00005002));
+ *	mmio_write_32(SYSCFG_BASE + SYSCFG_ICNVDEBWLCR, U(0x00003502));
+ */
+}
-- 
2.47.2

